homeassistant/components/backup/util.py

```diff
@@ -106,6 +123,70 @@ def validate_password(path: Path, password: str | None) -> bool:
     return False
 
 
+class AsyncIteratorReader:
+    """Wrap an AsyncIterator."""
+
+    def __init__(self, hass: HomeAssistant, stream: AsyncIterator[bytes]) -> None:
+        """Initialize the wrapper."""
+        self._hass = hass
+        self._stream = stream
+        self._buffer: bytes | None = None
+        self._pos: int = 0
+
+    async def _next(self) -> bytes | None:
+        """Get the next chunk from the iterator."""
+        return await anext(self._stream, None)
+
+    def read(self, n: int = -1, /) -> bytes:
+        """Read data from the iterator."""
+        result = bytearray()
+        while n < 0 or len(result) < n:
+            if not self._buffer:
+                self._buffer = asyncio.run_coroutine_threadsafe(
+                    self._next(), self._hass.loop
+                ).result()
+                self._pos = 0
+            if not self._buffer:
+                # The stream is exhausted
+                break
+            chunk = self._buffer[self._pos : self._pos + n]
+            result.extend(chunk)
+            n -= len(chunk)
+            self._pos += len(chunk)
+            if self._pos == len(self._buffer):
+                self._buffer = None
+        return bytes(result)
+
+
+def validate_password_stream(
+    input_stream: IO[bytes],
+    password: str | None,
+) -> None:
+    """Decrypt a backup."""
+    with (
+        tarfile.open(fileobj=input_stream, mode="r|", bufsize=BUF_SIZE) as input_tar,
+    ):
+        for obj in input_tar:
+            if not obj.name.endswith((".tar", ".tgz", ".tar.gz")):
+                continue
+            if obj.pax_headers.get(VERSION_HEADER) != "2.0":
+                raise UnsuppertedSecureTarVersion
+            istf = SecureTarFile(
+                None,  # Not used
+                gzip=False,
+                key=password_to_key(password) if password is not None else None,
+                mode="r",
+                fileobj=input_tar.extractfile(obj),
+            )
+            with istf.decrypt(obj) as decrypted:
+                try:
+                    decrypted.read(1)  # Read a single byte to trigger the decryption
+                except SecureTarReadError as err:
+                    raise IncorrectPassword from err
+                return
+    raise BackupEmpty
```

Comment:
Should we pass a message argument to the exception here? We're logging it in the manager now.

Reply:
I think it would be better to make these classes have default messages, I'll add that 👍